### PR TITLE
feat: implement bytebuffer

### PR DESCRIPTION
**Pull request by tisonkun**

### CODE REVIEW

File: kafka-api/src/bytebuffer/mod.rs [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+struct Shared {
+    ptr: *mut u8,
+}
+
+unsafe impl Send for Shared {}
+unsafe impl Sync for Shared {}
+
+impl Drop for Shared {
+    fn drop(&mut self) {
+        unsafe { drop_in_place(self.ptr) }
```
Comment (kezhuw) [inline kafka-api/src/bytebuffer/mod.rs:45]: This sounds like a memory leak.

I think  [`drop_in_place`](https://doc.rust-lang.org/std/ptr/fn.drop_in_place.html) is `unsafe` of [`Drop::drop`](https://doc.rust-lang.org/std/ops/trait.Drop.html) which [You cannot call Drop::drop yourself](https://doc.rust-lang.org/std/ops/trait.Drop.html#you-cannot-call-dropdrop-yourself) while [`drop`](https://doc.rust-lang.org/std/mem/fn.drop.html) consumes `self`. That is it does nothing with memory, but only deconstructor.

Comment (tisonkun) [reply kafka-api/src/bytebuffer/mod.rs:45]: How can I drop the underlying array then?

Comment (tisonkun) [reply kafka-api/src/bytebuffer/mod.rs:45]:
```patch
diff --git a/kafka-api/src/bytebuffer/mod.rs b/kafka-api/src/bytebuffer/mod.rs
index b81d090..d803430 100644
--- a/kafka-api/src/bytebuffer/mod.rs
+++ b/kafka-api/src/bytebuffer/mod.rs
@@ -13,7 +13,7 @@
 // limitations under the License.
 
 use core::slice;
-use std::{fmt::Debug, mem::ManuallyDrop, ops::RangeBounds, ptr::drop_in_place, sync::Arc};
+use std::{fmt::Debug, mem::ManuallyDrop, ops::RangeBounds, sync::Arc};
 
 use bytes::Buf;
 
@@ -24,6 +24,7 @@ mod impl_traits;
 #[derive(Debug, Clone)]
 struct Shared {
     ptr: *mut u8,
+    len: usize,
 }
 
 // shared ptr never changed + no alloc + in place mutations are managed
@@ -32,7 +33,7 @@ unsafe impl Sync for Shared {}
 
 impl Drop for Shared {
     fn drop(&mut self) {
-        unsafe { drop_in_place(self.ptr) }
+        unsafe { drop(Box::from_raw(slice::from_raw_parts_mut(self.ptr, self.len))) }
     }
 }
 
@@ -68,7 +69,7 @@ impl ByteBuffer {
         let mut me = ManuallyDrop::new(v);
         let (ptr, end) = (me.as_mut_ptr(), me.len());
         let start = 0;
-        let shared = Arc::new(Shared { ptr });
+        let shared = Arc::new(Shared { ptr, len: end });
         ByteBuffer { start, end, shared }
     }
```

Perhaps this one can work?

Comment (kezhuw) [reply kafka-api/src/bytebuffer/mod.rs:45]: How about [Vec::from_raw_parts](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.from_raw_parts) ? Or I can try [std::alloc::dealloc](https://doc.rust-lang.org/std/alloc/fn.dealloc.html) ? Both need `capacity`. There is no [`malloc`](https://en.cppreference.com/w/c/memory/malloc) counterpart, which need only `ptr` in rust.

Comment (kezhuw) [reply kafka-api/src/bytebuffer/mod.rs:45]: No sure. But I think [Vec::from_raw_parts](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.from_raw_parts) sound safer 😵‍💫 .

Comment (tisonkun) [reply kafka-api/src/bytebuffer/mod.rs:45]: Fixed alongside https://github.com/tisonkun/kafka-api/pull/5/files#diff-dab1a36778967a3dcfae0174f9f25d83f847abbc78678abd955bd2267bdfefc6. PTAL.

**Pull request merged**